src/erlcloud_aws.erl

```diff
@@ -796,14 +797,55 @@ service_host( <<"s3">>, Region ) ->
 service_host( <<"iam">>, <<"cn-north-1">> ) -> "iam.amazonaws.com.cn";
 service_host( <<"iam">>, <<"cn-northwest-1">> ) -> "iam.amazonaws.com.cn";
 service_host( <<"sdb">>, <<"us-east-1">> ) -> "sdb.amazonaws.com";
-service_host( <<"states">>, Region ) ->
-    binary_to_list( <<"states.", Region/binary, ".amazonaws.com">> );
 service_host( Service, <<"cn-north-1">> = Region ) when is_binary(Service) ->
     binary_to_list( <<Service/binary, $., Region/binary, ".amazonaws.com.cn">> );
-service_host( Service, <<"cn-northwest-1">> =Region ) when is_binary(Service) ->
+service_host( Service, <<"cn-northwest-1">> = Region ) when is_binary(Service) ->
     binary_to_list( <<Service/binary, $., Region/binary, ".amazonaws.com.cn">> );
-service_host( Service, Region ) when is_binary(Service) ->
-    binary_to_list( <<Service/binary, $., Region/binary, ".amazonaws.com">> ).
+% services which can have VPCe configured to mitigate cross-AZ traffic.
+% It's application level decision to use VPCe and configure those
+% magic can be done vi EC2 DescribeVpcEndpoints/filter by VPC/filter by AZ,
+% however permissions and describe* API throttling is not what we want to deal with here.
+service_host( Service, Region ) when is_binary(Service) andalso is_binary(Region) ->
+    Default = binary_to_list( <<Service/binary, $., Region/binary, ".amazonaws.com">> ),
+    get_host_vpc_endpoint(Service, Default).
+
+-spec get_host_vpc_endpoint(binary(), binary()) -> binary().
+%% take the list of possibly configured endpoints
+%% and pick the one which suites from our AZ.
+get_host_vpc_endpoint(Service, Default) when is_binary(Service) ->
+    ConfiguredEndpoints = proplists:get_value(Service,
+        application:get_env(erlcloud, services_vpc_endpoints, [])),
+    %% resolve through ENV if any
+    Endpoints = case ConfiguredEndpoints of
+        {env, EnvVarName} when is_list(EnvVarName) ->
+            Es = string:split(os:getenv(EnvVarName, ""), ",", all),
+            [list_to_binary(E) || E <- Es];
+        EndpointsList when is_list(EndpointsList) ->
+            EndpointsList
+    end,
+    % now get our AZ and match
+    case erlcloud_ec2_meta:get_instance_metadata("placement/availability-zone", default_config()) of
```

Comment:
let's avoid calling instance metadata if endpoints are not configured.

Reply:
Is this worth caching? I feel like it should never change without an application restart, is intrinsically application-wide, and therefore could be encapsulated into an ephemeral application env setting. Do we need to spam and parse instance metadata every time we want the service host?